Add new information about recent CSS inheritance changes with CSS selection and highlight pseudo-classes (#38989)
mdn/content: files/en-us/web/css/css_pseudo-elements/index.md

Pseudo-elements enable targeting entities not included in HTML and sections of content that cannot be targeted otherwise without adding extra markup. Consider the placeholder of an {{HTMLelement("input")}} element. This is an abstract element and not a distinct node in the document tree. You can select this placeholder by using the {{CSSXref("::placeholder")}} pseudo-element. As another example, the {{CSSXref("::selection")}} pseudo-element matches the content currently highlighted by a user, allowing you to style what is matched as the user interacts with the content and changes the selection. Similarly, the {{CSSXref("::first-line")}} pseudo-element targets the first line of an element, updating automatically if the character count of the first line changes, without having to query the element's line length.

> ## Description

Reviewer: OK, the biggest issue here is something I've only just noticed (sorry about that) — this content is on the wrong pseudo-elements page. This page is about the pseudo-elements module, and module pages just list information about is contained in that module. They don't include details like this. So, the first thing you should do is move this content to https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements — this is the main pseudo-elements landing page (but pseudo-elements, the concept, not pseudo-elements, the module).

Second thing — get rid of the "Description" heading, and promote "Highlight pseudo-elements inheritance" to an H2 (and promote the subheadings up a level).

I'd place the section on https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements, just above the "Specifications" section.

### Highlight pseudo-elements inheritance

> Highlight pseudo-elements, such as {{CSSxref("::selection")}}, {{CSSxref("::target-text")}}, {{CSSxref("::highlight()")}}, {{CSSxref("::spelling-error")}}, and {{CSSxref("::grammar-error")}}, follow a consistent inheritance model that differs from regular element inheritance.

Reviewer: ```suggestion
Highlight pseudo-elements, such as {{CSSxref("::selection")}}, {{CSSxref("::target-text")}}, {{CSSxref("::highlight()")}}, {{CSSxref("::spelling-error")}}, and {{CSSxref("::grammar-error")}}, follow a consistent inheritance model that differs from [regular element inheritance](/en-US/docs/Web/CSS/CSS_cascade/Inheritance).
```

When you apply styles to highlight pseudo-elements, they inherit from both:

> 1. Their parent elements (following normal inheritance)

Reviewer: ```suggestion
1. Their parent elements (following normal inheritance).
```

> 2. The highlight pseudo-elements of their parent elements (following highlight inheritance)

Reviewer: ```suggestion
2. The highlight pseudo-elements of their parent elements (following highlight inheritance).
```

This means that if you style both a parent element's highlight pseudo-element and a child element's highlight pseudo-element, the child's highlighted text will combine properties from both sources.

> 

Reviewer: I would move the content from the "Example" section here, as it is an example that relates to Highlight pseudo-elements inheritance behavior in general, not specific to custom props or universal selector behavior.

> #### Custom Properties Behavior

Reviewer: ```suggestion
#### Custom properties behavior
```

> CSS custom properties (variables) in highlight pseudo-elements inherit from their originating element (the element they're being applied to), not through the highlight inheritance chain. For example:

Reviewer: ```suggestion
[CSS custom properties (variables)](/en-US/docs/Web/CSS/CSS_cascading_variables/Using_CSS_custom_properties) in highlight pseudo-elements inherit from their originating element (the element they're being applied to), not through the highlight inheritance chain. For example:
```

```css

> :root {

Reviewer: This example could do with a little bit of explanation to describe what it is showing. At present, it isn't very clear. Maybe also include an HTML snippet that it could be applied to?

  --selection-color: lightgreen;
}

::selection {
  color: var(--selection-color);
}

.blue {
  --selection-color: blue;
}
```

> #### Universal Selector Considerations

Reviewer: ```suggestion
#### Universal selector considerations
```

When using the universal selector with highlight pseudo-elements, it prevents highlight inheritance. For example:

```css
/* This prevents highlight inheritance */
*::selection {
  color: lightgreen;
}

/* Prefer this to allow inheritance */
:root::selection {
  color: lightgreen;
}
```

> #### Example

Reviewer: When you move this content up, as suggested above, don't include the "Example" heading.

> This example demonstrates how highlight pseudo-elements inherit styles from their parent elements.

Reviewer: ```suggestion
Let's look at an example to demonstrate how highlight pseudo-elements inherit styles from their parent elements.

First, we have some HTML that includes two nested {{htmlelement("div")}} elements. Some of the included text content is contained directly inside the parent `<div>`, and some is nested inside the child `<div>`.
```

```html
<div class="parent">
  Parent text
  <div class="child">Child text</div>
</div>
```

> 

Reviewer: ```suggestion

Next we include some CSS, which selects the parent and child `<div>` elements separately and gives them different {{cssxref("color")}} values, and selects the parent and child's selected text ({{cssxref("::selection")}}), giving each a different {{cssxref("background-color")}} as well as setting a different text `color` on the parent selection.

```

```css
/* Style for the parent element */
.parent {
  color: blue;
}

/* Style for the parent's selected text */
.parent::selection {
  background-color: yellow;
  color: red;
}

/* Style for the child element */
.child {
  color: green;
}

/* Style for the child's selected text */
.child::selection {
  background-color: orange;
}
```

> Try selecting the text in both the parent and child elements. Notice that:
> 
> 1. When you select the parent text, it uses the yellow background and red text color defined in `.parent::selection`
> 2. When you select the child text, it uses:
>    - The orange background from `.child::selection`
>    - The red text color inherited from the parent's `::selection` pseudo-element
>    - This demonstrates how the child's highlight pseudo-element inherits from both its parent element and the parent's highlight pseudo-element
> 
> {{EmbedLiveSample("Example", "100%", "150")}}

Reviewer: ```suggestion
The example renders as follows:

{{EmbedLiveSample("Example", "100%", "150")}}

Try selecting the text in both the parent and child elements. Notice that:

1. When you select the parent text, it uses the yellow background and red text color defined in `.parent::selection`.
2. When you select the child text, it uses:
   - The orange background from `.child::selection`.
   - The red text color inherited from the parent's `::selection` pseudo-element.

This demonstrates how the child's highlight pseudo-element inherits from both its parent element and the parent's highlight pseudo-element.
```

> [!NOTE]

> > The exact appearance of selected text may vary slightly between browsers. For detailed information about browser support and implementation details, see the [CSS Custom Highlight API](/en-US/docs/Web/API/CSS_Custom_Highlight_API) documentation.

Reviewer: I still don't understand how this note helps us find information about support for the different highlight pseudo-elements. I mean, it lists `::highlight()` information, but what about the others? I'd suggest just removing it, or saying that you should consult the reference pages for the different pseudo-classes.

Pseudo-elements enable targeting entities not included in HTML and sections of content that cannot be targeted otherwise without adding extra markup. Consider the placeholder of an {{HTMLelement("input")}} element. This is an abstract element and not a distinct node in the document tree. You can select this placeholder by using the {{CSSXref("::placeholder")}} pseudo-element. As another example, the {{CSSXref("::selection")}} pseudo-element matches the content currently highlighted by a user, allowing you to style what is matched as the user interacts with the content and changes the selection. Similarly, the {{CSSXref("::first-line")}} pseudo-element targets the first line of an element, updating automatically if the character count of the first line changes, without having to query the element's line length.

## Highlight pseudo-elements inheritance

Highlight pseudo-elements, such as {{CSSxref("::selection")}}, {{CSSxref("::target-text")}}, {{CSSxref("::highlight()")}}, {{CSSxref("::spelling-error")}}, and {{CSSxref("::grammar-error")}}, follow a consistent inheritance model that differs somewhat from regular element inheritance.

> Styles applied to highlight pseudo-elements inherit from their parent elements. This means that if you apply a style to a parent element and then to the highlight pseudo-element of its child, the latter will inherit from both its parent element and the highlight pseudo-element styles of its parent.

Reviewer: I'm finding it hard to understand the inheritance model as described. Maybe it would be clearer if you added a quick code example to explain what you mean. You could show the live rendered result using the EmedLiveSample macro; see https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Live_samples for an explanation.

> For detailed information and examples of how this inheritance model works in practice, refer to the [CSS Custom Highlight API](/en-US/docs/Web/API/CSS_Custom_Highlight_API) documentation.

Reviewer: Looking at the API landing page, I can't see any explanation of the inheritance model. Is it in a particular subpage?

BTW, you could always link to the Chrome page that you linked to previously for the detailed explanation if it is not already on MDN. We just need some information on MDN as well.
